drivers/gpio/gpio_pca_series.c

```diff
@@ -0,0 +1,2077 @@
+						   enum gpio_pca_series_reg_type reg_type)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
+
+	return cfg->part_cfg->regs[reg_type];
+}
+
+/**
+ * @brief get read size for register
+ *
+ * @param dev
+ * @param reg_type
+ * @return uint32_t size in bytes
+ *                  0 if fail
+ */
+static inline uint32_t gpio_pca_series_reg_get_size(const struct device *dev,
+							enum gpio_pca_series_reg_type reg_type)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
+
+	switch (reg_type) {
+	case gpio_pca_series_1b_input_port:
+	case gpio_pca_series_1b_output_port:
+/*  case gpio_pca_series_1b_polarity_inversion: */
+	case gpio_pca_series_1b_configuration:
+	case gpio_pca_series_1b_input_latch:
+	case gpio_pca_series_1b_pull_enable:
+	case gpio_pca_series_1b_pull_select:
+	case gpio_pca_series_1b_input_status:
+	case gpio_pca_series_1b_output_config:
+#ifdef CONFIG_GPIO_PCA_SERIES_INTERRUPT
+	case gpio_pca_series_1b_interrupt_mask:
+	case gpio_pca_series_1b_interrupt_status:
+	case gpio_pca_series_1b_interrupt_clear:
+#endif /* CONFIG_GPIO_PCA_SERIES_INTERRUPT */
+		return cfg->part_cfg->port_no;
+	case gpio_pca_series_2b_output_drive_strength:
+#ifdef CONFIG_GPIO_PCA_SERIES_INTERRUPT
+	case gpio_pca_series_2b_interrupt_edge:
+#endif /* CONFIG_GPIO_PCA_SERIES_INTERRUPT */
+		return (cfg->part_cfg->port_no * 2U);
+	default:
+		return 0; /** should never happen */
+	}
+}
+
+#ifdef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+/** forward declarations */
+static inline int gpio_pca_series_reg_update_cache(struct device *dev,
+						   enum gpio_pca_series_reg_type reg_type,
+						   uint8_t *buf);
+#endif /* CONFIG_GPIO_PCA_SERIES_CACHE_ALL */
+
+/**
+ * @brief read register with i2c interface.
+ *
+ * @note if CONFIG_GPIO_PCA_SERIES_CACHE_ALL is enabled, this will
+ *          also update reg cache.
+ *
+ * @param dev
+ * @param reg_type
+ * @param buf pointer to read data
+ * @return int 0        if success
+ *             -EFAULT  if register is not supported
+ *             -EIO     if i2c failure
+ */
+static inline int gpio_pca_series_reg_read(const struct device *dev,
+					   enum gpio_pca_series_reg_type reg_type, uint8_t *buf)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	/* struct gpio_pca_series_data *data = dev->data; */
+	int ret;
+	uint32_t size = gpio_pca_series_reg_get_size(dev, reg_type);
+	uint8_t addr = gpio_pca_series_reg_get_addr(dev, reg_type);
+	struct i2c_msg msg[2];
+
+	LOG_DBG("reg read type %d addr 0x%x len %d", reg_type, addr, size);
+
+	if (addr == PCA_REG_INVALID) {
+		LOG_ERR("trying to read unsupported reg, reg type %d", reg_type);
+		return -EFAULT;
+	}
+
+	msg[0].buf = &addr;
+	msg[0].len = 1;
+	msg[0].flags = I2C_MSG_WRITE;
+	msg[1].buf = buf;
+	msg[1].len = size;
+	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;
+
+	ret =  i2c_transfer(dev, msg, 2, addr);
```

Comment:
Seems like this could be replaced with a call to `i2c_reg_read_byte`?

Reply:
Not really. device could have multiple bytes in one "register", stored `size`, either because each pin requires 2 bits of configuration, or the device has more than 1 port, or both.
In #69330 aasinclair stated:
> Burst write is unreliable, so please use i2c_write instead (See https://github.com/zephyrproject-rtos/zephyr/issues/20154).

So I fallbacked to these code.